include/ceed.h

```diff
@@ -192,6 +192,29 @@ CEED_EXTERN int CeedSetErrorHandler(Ceed ceed,
 CEED_EXTERN int CeedGetErrorMessage(Ceed, const char **errmsg);
 CEED_EXTERN int CeedResetErrorMessage(Ceed, const char **errmsg);
 
+/// Ceed Errors
+///
+/// This enum is used to specify the type of error returned by a function.
+/// A zero error code is success, odd error codes indicate non-recoverable errors
+/// and even error codes indicate recoverable errors.
+/// @ingroup Ceed
+typedef enum {
+  /// Sucess error code
+  CEED_ERROR_SUCCESS       = 0,
+  /// Recoverable error, generic
+  CEED_ERROR_RECOVERABLE   = 2,
+  /// Recoverable error, dimension mismatch in inputs
+  CEED_ERROR_DIMENSION     = 4,
+  /// Recoverable error, incomplete object setup
+  CEED_ERROR_INCOMPLETE    = 6,
+  /// Unrecoverable error, generic
+  CEED_ERROR_UNRECOVERABLE = 1,
+  /// Unrecoverable error, internal backend error
+  CEED_ERROR_BACKEND       = 3,
+  /// Unrecoverable error, operation unsupported by current backend
+  CEED_ERROR_UNSUPPORTED   = 5,
+} CeedErrorType;
```

Comment:
@jedbrown, any opinions on the error codes?

Reply:
Any reason not to use positive and negative numbers instead of even/odd?

I'm not sure "recoverable" is the best term. It means that some object state might be inconsistent and/or resources may have leaked. "Recoverable" at this point should probably mean that the library and object states have not been modified since the call. We'll eventually want to distinguish incomplete modification of an output parameter, in which case the user could try again with different inputs (e.g., shorten a time step and retry).

Reply:
The only reason to use even/odd over positive/negative is to provide human readable error code strings like we do with our other enums.

Agreed re recoverable, but I'm not sure what is a good replacement word? How about terminal and nonterminal?

Reply:
The pointer can be shifted so there's no problem with an array of error strings.